be/src/exec/spill/input_stream.cpp

```diff
@@ -0,0 +1,242 @@
+// Copyright 2021-present StarRocks, Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "exec/spill/input_stream.h"
+
+#include "runtime/sorted_chunks_merger.h"
+#include "util/blocking_queue.hpp"
+#include "util/defer_op.h"
+
+namespace starrocks {
+namespace spill {
+
+static const int chunk_buffer_max_size = 2;
+
+class BufferedInputStream : public InputStream {
+public:
+    BufferedInputStream(int capacity, InputStreamPtr stream) : _capacity(capacity), _input_stream(stream) {}
+    ~BufferedInputStream() override = default;
+
+    bool is_buffer_full() { return _chunk_buffer.get_size() == _capacity || eof(); }
+    bool has_chunk() { return !_chunk_buffer.empty() || eof(); }
+
+    StatusOr<ChunkUniquePtr> get_next(FormatterContext& ctx) override;
+    bool is_ready() override { return has_chunk(); }
+    void close() override {}
+
+    bool enable_prefetch() const override { return true; }
+
+    Status prefetch(FormatterContext& ctx) override;
+
+    StatusOr<ChunkUniquePtr> read_from_buffer();
+
+private:
+    bool _acquire() {
+        bool expected = false;
+        return _is_prefetching.compare_exchange_strong(expected, true);
+    }
+    void _release() { _is_prefetching = false; }
```

Comment:
why don't use `compare_exchange_strong` ?

Reply:
> why don't use `compare_exchange_strong` ?

_release will be only invoked after _acquire successes, I think it's unnecessary to use cas here.

Reply:
would require and release be in race case?

Reply:
> would require and release be in race case?

maybe. multiple _acquire may be in race case, so we need CAS to ensure only one thread can go ahead. Once the CAS is successful, only this thread will invoke _release and _is_prefetching must be true before invoking.

Reply:
```        
void _release() {
        bool expected = true;
        return _is_prefetching.compare_exchange_strong(expected, false);
}
```
suggest, avoid _release() by mistake if _acquire falied.

Reply:
> ```
> void _release() {
>         bool expected = true;
>         return _is_prefetching.compare_exchange_strong(expected, false);
> }
> ```
> 
> suggest, avoid _release() by mistake if _acquire falied.

ok, I add a DCHECK here